lib/Config/Hooks.php

```diff
@@ -3,9 +3,17 @@
 namespace MailPoet\Config;
 
 use MailPoet\Models\Setting;
-use MailPoet\WP\Posts as WPPosts;
+use MailPoetVendor\Psr\Container\ContainerInterface;
 
 class Hooks {
+
+  /** @var ContainerInterface */
+  private $container;
+
+  function __construct(ContainerInterface $container) {
```

Comment:
Why do we inject the whole container here, not just the necessary class?

Reply:
Good question. Well I was hesitating here. The thing is that hooks are like events and they might not happen. So I did not want to instantiate all services for handling those actions every time. Symfony has a nice solution for this [a lazy services](https://symfony.com/doc/current/service_container/lazy_services.html) but it requires adding more dependencies (`symfony/proxy-manager-bridge` which it self has more deps).  I’ve chosen a way of injecting a container and then the service is requested from container after the action is triggered

Reply:
@costasovo I don't like that reasoning very much. You can say that pretty much about any functionality in the whole application. The container itself should not be injected (except maybe a router/endpoint-factory). Does the class require any really complicated dependency (i.e. dependency that reads from disk or does SQL queries during its initialization)? If not, then it's most likely premature optimization. Also, if a dependency really has "heavy" initialization then it's most likely a badly designed dependency. I don't like that just after introducing DI we are already breaking it and using it as a service locator.

Reply:
I agree that injecting container is not good thing and I want to avoid it. I just think that these hooks are very similar to a router.

Reply:
@JanJakes I agree with you, the `\MailPoet\Subscription\Form` initialization doesn't look heavy aside from a usual memory footprint of a loaded class. So we can abstain from lazy loading here until it really becomes a performance bottleneck.

Reply:
But the currently required service doesn't need any complicated deps so I can inject it directly and maybe do some optimizations later if necessary.

Reply:
Ok then I'll inject the service. 👍